### PR TITLE
INT-3663: Fix early BF access for Security Module

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3663

Previously the `ChannelSecurityInterceptorBeanPostProcessor` was populated with direct `BeanDefinition`s for `ChannelSecurityInterceptor`s.
It caused an `early access to BeanFactory`.
The issue has been introduced by the `ChannelSecurityInterceptorFactoryBean`
- Rework `SecurityIntegrationConfigurationInitializer` do not populate `BeanDefinition`s to the `ChannelSecurityInterceptorBeanPostProcessor`, but just `bean names`
- Redesign `ChannelSecurityInterceptorBeanPostProcessor` to the `AbstractAutoProxyCreator`
- Introduce `SecuredChannel` annotation to be used on the `@Bean` level for `MessageChannel` definition
- Move `access policy` mapping to the `SecuredChannel` annotation

### CODE REVIEW

**Comment by artembilan** [comment]
Need some polishing, JavaDocs and Docs.
And Migration Guide note afterwards.

**Comment by artembilan** [comment]
@garyrussell , let me know if you are OK with the solution and I go ahead for docs.

File: spring-integration-security/src/main/java/org/springframework/integration/security/config/ChannelSecurityInterceptorBeanPostProcessor.java [hunk 1]
```diff
@@ -37,40 +40,60 @@
  * @author Oleg Zhurakousky
  * @author Artem Bilan
  */
-public class ChannelSecurityInterceptorBeanPostProcessor implements BeanPostProcessor {
+public class ChannelSecurityInterceptorBeanPostProcessor extends AbstractAutoProxyCreator {
```
Comment (garyrussell) [inline spring-integration-security/src/main/java/org/springframework/integration/security/config/ChannelSecurityInterceptorBeanPostProcessor.java]: WARN: serialVersionUID

File: spring-integration-security/src/main/java/org/springframework/integration/security/channel/DefaultChannelAccessPolicy.java [hunk 2]
```diff
@@ -72,7 +72,40 @@ public DefaultChannelAccessPolicy(String sendAccess, String receiveAccess) {
...
+	 * Typically is used for the values from the {@link SecuredChannel}
+	 * @param sendAccess The send access values.
+	 * @param receiveAccess The receive access values.
+	 * @since 4.2
+	 */
+	public DefaultChannelAccessPolicy(String[] sendAccess, String[] receiveAccess) {
+		boolean sendAccessDefined = !ObjectUtils.isEmpty(sendAccess);
+		boolean receiveAccessDefined = !ObjectUtils.isEmpty(receiveAccess);
+		Assert.isTrue(sendAccessDefined || receiveAccessDefined,
+				"At least one of 'sendAccess' and 'receiveAccess' must not be null.");
```
Comment (garyrussell) [inline spring-integration-security/src/main/java/org/springframework/integration/security/channel/DefaultChannelAccessPolicy.java]: `... must not be null and have at least one entry`

**Comment by garyrussell** [comment]
Otherwise LGTM

**Comment by artembilan** [comment]
Pushed

**Comment by garyrussell** [comment]
LGTM; still needs docs etc, though; right?

**Comment by artembilan** [comment]
Yeah... Sure! At least say that we did a deprecation and add a couple samples with JavaConfig

**Comment by artembilan** [comment]
Pushed documentation.
After the merge I'll go ahead with Migration Guide note.

**Comment by garyrussell** [comment]
Merged as b6cfd4fa76cb39838fccec4c963d73dbb28b93b2

**Pull request closed without merging**